File: services/api/app/schemas/clima.py
"""Schemas Pydantic para M7 — Termómetro de clima laboral."""
from __future__ import annotations

from datetime import date, datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field


# ── Encuestas ─────────────────────────────────────────────────────────────────

# Dimensiones FPSICO/INSST (Factor Psicosocial — método del INSST)
DimensionFPSICO = Literal[
    "autonomia",
    "carga_trabajo",
    "demandas_cognitivas",
    "variedad_contenido",
    "participacion",
    "interes_compensacion",
    "desempeño_rol",
    "relaciones_apoyo",
]


class PreguntaClima(BaseModel):
    id: str
    texto: str
    dimension: DimensionFPSICO
    escala: Literal["likert5", "likert7", "binaria"] = "likert5"


class EncuestaActiva(BaseModel):
    id: UUID
    titulo: str
    descripcion: str | None
    fecha_inicio: date
    fecha_fin: date
    preguntas: list[PreguntaClima]


class CrearEncuestaRequest(BaseModel):
    titulo: str = Field(..., min_length=5, max_length=200)
    descripcion: str | None = None
    fecha_inicio: date
    fecha_fin: date
    dimensiones: list[DimensionFPSICO] = Field(..., min_length=1)


# ── Respuestas (totalmente anónimas — sin usuario_id) ─────────────────────────

class RespuestaItem(BaseModel):
    pregunta_id: str
    valor: int = Field(..., ge=1, le=7)  # escala 1–5 o 1–7


class EnviarRespuestasRequest(BaseModel):
    encuesta_id: UUID
    departamento: str | None = None   # solo categoría, nunca nombre
    respuestas: list[RespuestaItem] = Field(..., min_length=1)


# ── Indicadores agregados ─────────────────────────────────────────────────────

class IndicadorDimension(BaseModel):
    dimension: DimensionFPSICO
    media: float              # 1.0–7.0
    mediana: float
    desviacion: float
    n_respuestas: int         # nunca < K_ANONIMATO_MIN
    nivel_riesgo: Literal["bajo", "medio", "alto"]
    variacion_pct: float | None  # % vs. periodo anterior


class IndicadoresClima(BaseModel):
    encuesta_id: UUID
    titulo: str
    fecha_inicio: date
    fecha_fin: date
    n_total_respuestas: int
    indice_global: float       # media ponderada 1–7
    suficiente_anonimato: bool  # False si n < K_ANONIMATO_MIN
    dimensiones: list[IndicadorDimension]
    alertas: list[str]          # p. ej. ["relaciones_apoyo en riesgo alto"]

File: packages/shared-types/src/clima.ts
// ============================================================
// SafeWork AI — Tipos del M7: Termómetro de clima laboral
//               y M6: Reporting / Dashboard
// ============================================================

// --- Enums clima (M7) ---

// Dimensiones del modelo FPSICO / INSST
export type DimensionFpsico =
  | "carga_trabajo"
  | "autonomia"
  | "apoyo_social"
  | "relaciones"
  | "conflicto_rol"
  | "participacion"
  | "interes_compensacion"
  | "desempeno_rol";

export type NivelRiesgoClimaLaboral =
  | "muy_adecuado"    // índice ≥ 80
  | "adecuado"        // índice 60–79
  | "moderado"        // índice 40–59
  | "elevado"         // índice 20–39
  | "muy_elevado";    // índice < 20

// --- Encuesta (M7) ---

export interface PreguntaClima {
  id: string;
  encuestaId: string;
  orden: number;
  enunciado: string;
  dimension: DimensionFpsico;
  escala: "likert_5" | "likert_7" | "binaria";
}

export interface EncuestaClima {
  id: string;
  empresaId: string;
  mes: string;                          // "2026-05" — siempre primer día del mes
  activa: boolean;
  preguntas: PreguntaClima[];
  departamentos: string[];              // a qué departamentos se envía
  fechaLimiteRespuesta: string;
  numRespuestasMinimas: number;         // k-anonimato: mínimo para mostrar resultados
  createdAt: string;
}

// La respuesta NO tiene FK a usuario — anonimato real
export interface RespuestaClima {
  id: string;
  encuestaId: string;
  departamento: string;                 // solo agregado por departamento
  respuestas: Record<string, number>;   // preguntaId → valor numérico
  respondidaAt: string;
  // Sin usuarioId, sin IP, sin ningún identificador
}

// --- Indicadores agregados (M7) ---

export interface IndicadorDimension {
  dimension: DimensionFpsico;
  puntuacion: number;                   // 0-100
  nivelRiesgo: NivelRiesgoClimaLaboral;
  variacionMesAnterior?: number;        // +/- puntos
}

export interface IndicadorClima {
  encuestaId: string;
  empresaId: string;
  mes: string;
  departamento?: string;               // null = empresa completa
  numRespuestas: number;
  satisfaceKAnonimato: boolean;        // false → no se muestran resultados
  indiceGlobal: number;               // 0-100
  nivelRiesgoGlobal: NivelRiesgoClimaLaboral;
  dimensiones: IndicadorDimension[];
  alertaEnviada: boolean;
  tendencia: "mejora" | "estable" | "deterioro" | "sin_datos";
  calculadoAt: string;
}

// --- Dashboard ejecutivo (M6) ---

export interface KpisDashboard {
  periodo: string;                      // "2026-Q1", "2026-05", "2026"
  // Canal de denuncia
  totalDenuncias: number;
  denunciasPorEstado: Record<string, number>;
  tiempoMedioResolucionDias: number;
  // Formación
  coberturaFormativa: number;           // % de empleados con al menos 1 curso completado
  certificadosEmitidos: number;
  // Clima
  indiceClimaGlobal?: number;
  departamentosEnRiesgo: number;
  // Cumplimiento
  protocoloVigente: boolean;
  dpaFirmado: boolean;
  formacionObligatoriaCompletada: boolean;
}

// Informe del Plan de Igualdad (sección de acoso)
export interface InformePlanIgualdad {
  empresaId: string;
  anio: number;
  fechaGeneracion: string;
  kpis: KpisDashboard;
  medidasAdoptadas: string[];           // lista de medidas del protocolo
  indicadoresClimaAnuales: IndicadorClima[];
  certificadosFormacion: number;
  incidencias: number;
  resolucionesFavorables: number;
  tiempoMedioResolucionDias: number;
  firmaResponsable?: string;
}
